relayer-engine/src/executor/executorHarness.ts

```diff
@@ -75,16 +76,38 @@ async function spawnExecutor(
   plugins: Plugin[],
   providers: Providers,
   workerInfoMap: Map<ChainId, WorkerInfo[]>,
-  logger: ScopedLogger
+  logger: ScopedLogger,
 ): Promise<void> {
   const actionQueues = spawnWalletWorkers(providers, workerInfoMap);
 
   while (true) {
-    await sleep(SPAWN_WORKFLOW_INTERNAL);
     try {
-      if ((await storage.numActiveWorkflows()) < MAX_ACTIVE_WORKFLOWS) {
-        await spawnWorkflow(storage, plugins, providers, actionQueues, logger);
+      if ((await storage.numActiveWorkflows()) >= MAX_ACTIVE_WORKFLOWS) {
+        await sleep(SPAWN_WORKFLOW_INTERNAL);
+        continue;
+      }
+      // TODO
+      const res = await storage.getNextWorkflow(1);
+      if (!res) {
+        logger.debug("No new workflows found");
+        continue;
       }
+      // After getting a workflow due to racing conditions we may be attempting to process more than MAX, if so, requeue and try again
```

Comment:
Thinking about this a bit more, the reason to limit the number of active workflows is essentially a performance throttling measure per process. The executor hasn't been designed to operate multi-node, but in that case the numActiveWorkflows should be per node, not global. 

In that case, since this loop should be unique per node and getNextWorkflow is awaited here, it should impossible to race.